### ISSUE TITLE
Setting and escaping a string containing the "_xHHHH_" pattern into a cell

### ISSUE THREAD

**Issue by EmmanuelBrunet** [ClosedXML/ClosedXML#608]
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

0.89.0

**What is the current behavior?**

When setting a string value containing the "_xHHHH_" pattern into a cell, this string value is written "as is" into the "sharedStrings.xml" file inside the Excel file. However, Excel interprets this pattern as an Unicode character.
Example : setting the "Reserve_TT_A_BLOCAGE_CAG_x6904_2" string into a cell results in Excel displaying "Reserve_TT_A_BLOCAGE_CAG椄2" ("椄" = Unicode character U+6904).

**What is the expected behavior or new feature?**

Using already the source code of ClosedXML to add a function that gets the colors of all characters of a RichText cell, I implemented a work-around for this problem, inspired by the usage of the "utfPattern" regular expression already used by "private static bool TryGetStringValue<T>(out T value, object currValue)" in the ".\ClosedXML\Excel\Cells\XLCell.cs" file.

In this same file, I changed the beginning of "public IXLCell SetValue<T>(T value)" into:

        public IXLCell SetValue<T>(T value)
        {
            if (value == null)
                return this.Clear(XLClearOptions.Contents);

            FormulaA1 = String.Empty;
            _richText = null;
            var style = GetStyleForRead();
            if (value is String || value is char)
            {
                string val = value.ToString();
                if (utfPattern.Match(val).Success)
                {
                    var sb = new StringBuilder();
                    var lastIndex = 0;
                    foreach (Match match in utfPattern.Matches(val))
                    {
                        var matchString = match.Value;
                        var matchIndex = match.Index;
                        sb.Append(val.Substring(lastIndex, matchIndex - lastIndex));
                        sb.Append("_x005F");
                        lastIndex = matchIndex;
                    }
                    if (lastIndex < val.Length)
                        sb.Append(val.Substring(lastIndex));
                    val = sb.ToString();
                }
                _cellValue = val;
                _dataType = XLCellValues.Text;
                if (_cellValue.Contains(Environment.NewLine) && !GetStyleForRead().Alignment.WrapText)
                    Style.Alignment.WrapText = true;
            }

And the beginning of "private void SetValue(object value)" into:

        private void SetValue(object value)
        {
            FormulaA1 = String.Empty;
            string val;
            if (value == null)
                val = string.Empty;
            else if (value is DateTime)
                val = ((DateTime)value).ToString("o");
            else if (value.IsNumber())
                val = Convert.ToDecimal(value).ToInvariantString();
            else
            {
                val = value.ToString();
                if (utfPattern.Match(val).Success)
                {
                    var sb = new StringBuilder();
                    var lastIndex = 0;
                    foreach (Match match in utfPattern.Matches(val))
                    {
                        var matchString = match.Value;
                        var matchIndex = match.Index;
                        sb.Append(val.Substring(lastIndex, matchIndex - lastIndex));
                        sb.Append("_x005F");
                        lastIndex = matchIndex;
                    }
                    if (lastIndex < val.Length)
                        sb.Append(val.Substring(lastIndex));
                    val = sb.ToString();
                }
            }

This work-around fills my needs by "escaping" the "_xHHHH_" pattern by adding a leading "_x005F" (the Unicode underscore character). This escape sequence now appears inside the "sharedStrings.xml" file and Excel correctly displays my string, without this escape sequence and without changing the "_x6904_" string into "椄".

The attached Excel file has been created with ClosedXML v0.89 + my work-around that adds the "_x005F" escape sequence.

**Did this work in previous versions of our tool?  Which versions?**

Yes/v0.89

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

No code sample is necessary, since I propose a work-around inside ClosedXML, based on a extremely rare problem. The software I'm working on manipulates large numbers of data, including variable names such as the one I described earlier: "Reserve_TT_A_BLOCAGE_CAG_x6904_2".

**Code to reproduce problem:**

None.

- [[Delta_Interface_v012_v011.xlsx](https://github.com/ClosedXML/ClosedXML/files/1514229/Delta_Interface_v012_v011.xlsx)] I attached a sample spreadsheet.  (You can drag files on to this issue)

**Comment by igitur** [comment]
You can test the build at https://ci.appveyor.com/project/Pyropace/closedxml/build/0.8.721/artifacts

You will have to install a new, additional dependency from NuGet: `ExcelNumberFormat`

**Comment by igitur** [comment]
PS: You're lucky I didn't close this issue for not submitting a code sample. I'm not sure what more in the template I can say to make it clear that it isn't optional. Here is a code sample that would replicate the issue:
```c#
        public void CanEscape_xHHHH_Correctly()
        {
            using (var ms = new MemoryStream())
            {
                using (var wb = new XLWorkbook())
                {
                    var ws = wb.AddWorksheet("Sheet1");
                    ws.FirstCell().Value = "Reserve_TT_A_BLOCAGE_CAG_x6904_2";
                    wb.SaveAs(ms);
                }

                ms.Seek(0, SeekOrigin.Begin);

                using (var wb = new XLWorkbook(ms))
                {
                    var ws = wb.Worksheets.First();
                    Assert.AreEqual("Reserve_TT_A_BLOCAGE_CAG_x6904_2", ws.FirstCell().Value);
                }
            }
        }
```

**Comment by EmmanuelBrunet** [comment]
Thank you very much.

However, I will stick to my modified ClosedXML v0.89.
It fits my needs very well, and the software version I work on is « critical ».
It features a lot of optimizations, reducing the generation times from hours into minutes, thanks to ClosedXML (replacing the Excel Interop) and to a lot of optimizations (file management, Dispose, and so on).

Maybe for a future usage of this fabulous library into another project, since the Excel Interop is used in several software, Excel files being a common way of managing input data.

Bye,
Emmanuel

De : Francois Botha [mailto:notifications@github.com]
Envoyé : mercredi 29 novembre 2017 16:59
À : ClosedXML/ClosedXML
Cc : Emmanuel BRUNET; Author
Objet : Re: [ClosedXML/ClosedXML] Setting and escaping a string containing the "_xHHHH_" pattern into a cell (#608)


You can test the build at https://ci.appveyor.com/project/Pyropace/closedxml/build/0.8.721/artifacts

You will have to install a new, additional dependency from NuGet: ExcelNumberFormat

**Comment by EmmanuelBrunet** [comment]
Thank you again, igitur.

As of the missing code sample, sorry about it.
I'm new to the gitHub community and I just wanted to warn you about a specific behaviour of Excel that was not implemented into ClosedXML.
I thought that it was not necessary to incorporate a generic code sample because I described the work-around I used for my own needs.
I just wanted to help you a bit to make ClosedXML better ^^.
